src/chess_widget.rs

```diff
@@ -61,28 +108,85 @@ impl ChessWidget {
 
     fn draw_square(&self, ctx: &mut PaintCtx, position: Position) {
         if let Some(piece) = self.game.get_piece(position) {
-            let image_data = get_image(self.get_image_file(piece));
             let image = ctx
                 .make_image(
                     PIECE_SIZE,
                     PIECE_SIZE,
-                    &image_data,
+                    self.get_image_file(piece),
                     ImageFormat::RgbaSeparate,
                 )
                 .unwrap();
-            let x0 = WINDOW_SIZE * position.x as f64 / 8.0;
-            let y1 = WINDOW_SIZE * (1.0 - (position.y as f64) / 8.0);
+            let mut p0 = Point::from(position);
+            // if we are holding a piece, offset it's position by how far it's been dragged
+            if let Some(mouse_down) = self.mouse_down {
+                if position == Position::from(mouse_down) {
+                    p0.x += self.current_point.x - mouse_down.x;
+                    p0.y += self.current_point.y - mouse_down.y;
+                }
+            }
             ctx.draw_image(
                 &image,
-                Rect::new(x0, y1 - WINDOW_SIZE / 8.0, x0 + WINDOW_SIZE / 8.0, y1),
+                Rect::new(
+                    p0.x,
+                    p0.y,
+                    p0.x + WINDOW_SIZE / 8.0,
+                    p0.y + WINDOW_SIZE / 8.0,
+                ),
                 InterpolationMode::Bilinear,
             );
         }
     }
+
+    // We want the square a dragged piece is considered to be on to be based on the center of
+    // the piece, not the location of the mouse. This requires offsetting based on the original
+    // mouse down location
+    fn get_dragged_piece_position(&self, mouse_up: Point) -> Position {
+        let mouse_down = self.mouse_down.unwrap();
+        let top_left = Point::from(Position::from(mouse_down));
+        let middle = Point {
+            x: top_left.x + WINDOW_SIZE / 16.0,
+            y: top_left.y + WINDOW_SIZE / 16.0,
+        };
+        let x_offset = mouse_down.x - middle.x;
+        let y_offset = mouse_down.y - middle.y;
+        Position::from(Point {
+            x: mouse_up.x - x_offset,
+            y: mouse_up.y - y_offset,
+        })
+    }
 }
 
 impl Widget<String> for ChessWidget {
-    fn event(&mut self, _ctx: &mut EventCtx, _event: &Event, _data: &mut String, _env: &Env) {}
+    fn event(&mut self, _ctx: &mut EventCtx, _event: &Event, _data: &mut String, _env: &Env) {
+        match _event {
```

Comment:
for `_event` the `_` usually means allow it to be unused without warning, but it's being used here in the match, so should probably remove right?

Reply:
Lol I didn't even know that, just copied the function definition. I'll change it.